### PR TITLE
chore: add pre-commit hook

### PR DESCRIPTION
**Pull request by Chisomchima**
Implements [DEVOPS-431](https://dhis2.atlassian.net/browse/DEVOPS-431)
---

#### Summary
This PR introduces git pre-commit hooks to ensure that code quality checks (formatting and linting) are executed automatically before any commits are made. This helps maintain code quality and consistency across the project.

#### Implementation Details
- **Husky**: Ensures that the pre-commit hook is triggered automatically when committing changes.
- **Pre-commit Hook Script**: Executes the following commands sequentially:
  - `yarn format`: Formats the codebase using Prettier.
  - `yarn lint`: Runs ESLint to catch any linting issues.

#### Checklist
- [x] Added Husky and configured pre-commit hooks.
- [x] Verified that the hooks run correctly and block commits on failures.

#### Notes for Reviewers
- Ensure that the commands run successfully on your local environment.
- Provide feedback on any additional checks or improvements that can be added to the pre-commit hook.

#### References
- [Husky Documentation](https://typicode.github.io/husky/#/)

### CODE REVIEW

**Comment by Mohammer5** [comment]
See https://dhis2.slack.com/archives/C078GG39MQU/p1721631184102679

File: .husky/pre-commit [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+yarn format && yarn lint
```
Comment (Mohammer5) [inline .husky/pre-commit:4]: `yarn format` should be enough. It exits with a non-zero code when there's code that can't be formatted

File: package.json [hunk 2]
```diff
@@ -31,7 +31,8 @@
         "format": "yarn d2-style apply",
         "lint": "yarn d2-style check",
         "e2e": "yarn playwright test --ui",
-        "generate-types": "rm -rf ./src/types/generated && yarn openapi --input https://dev.api.im.dhis2.org/swagger.yaml --output ./src/types/generated --client axios --useOptions --useUnionTypes --exportCore false --exportServices false --exportModels true --exportSchemas false && yarn prettier src/types/generated --write --loglevel silent"
+        "generate-types": "rm -rf ./src/types/generated && yarn openapi --input https://dev.api.im.dhis2.org/swagger.yaml --output ./src/types/generated --client axios --useOptions --useUnionTypes --exportCore false --exportServices false --exportModels true --exportSchemas false && yarn prettier src/types/generated --write --loglevel silent",
+        "prepare": "husky install"
```
Comment (Mohammer5) [inline package.json:35]: I know that this is the official recommendation, but I think it's easier to just do:

```suggestion
        "postinstall": "husky install"
```

This ensures it's always working, even when checking out the repo for the first time, without having to do these steps manually (and therefore having to know that you have to do these steps)

**Review by Mohammer5 — CHANGES_REQUESTED** [review]

File: .husky/pre-commit [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# format code style
+yarn format
+
+# handle linting errors
+ yarn lint
```
Comment (Mohammer5) [inline .husky/pre-commit:7]: `yarn format` should (will) exit with a non-zero code when there are remaining issues.

```suggestion
```

File: .husky/commit-msg [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+npx --no-install commitlint --edit "$1"
```
Comment (Mohammer5) [inline .husky/commit-msg:3]: Should be moved to `package.json` (`"lint-commit-msg": "npx --no-install commitlint --edit "$1""`).
Then this script becomes declarative (`yarn lint-commit-msg`) and all scripts are defined in a single location (`package.json`) and independent of the tools that use them

File: package.json [hunk 5]
```diff
@@ -31,7 +32,8 @@
         "format": "yarn d2-style apply",
         "lint": "yarn d2-style check",
         "e2e": "yarn playwright test --ui",
-        "generate-types": "rm -rf ./src/types/generated && yarn openapi --input https://dev.api.im.dhis2.org/swagger.yaml --output ./src/types/generated --client axios --useOptions --useUnionTypes --exportCore false --exportServices false --exportModels true --exportSchemas false && yarn prettier src/types/generated --write --loglevel silent"
+        "generate-types": "rm -rf ./src/types/generated && yarn openapi --input https://dev.api.im.dhis2.org/swagger.yaml --output ./src/types/generated --client axios --useOptions --useUnionTypes --exportCore false --exportServices false --exportModels true --exportSchemas false && yarn prettier src/types/generated --write --loglevel silent",
+        "prepare": "husky"
```
Comment (Mohammer5) [inline package.json:36]: See unresolved comment above (not sure why it's not showing in the code-tab anymore)

**Review by Mohammer5 — CHANGES_REQUESTED** [review]

File: .husky/pre-commit (hunk 3 above)
Comment (tonsV2) [reply .husky/pre-commit:7]: Isn't that a good thing? :slightly_smiling_face:

File: package.json (hunk 5 above)
Comment (Chisomchima) [reply package.json:36]: @Mohammer5  @tonsV2  I'll be closing this pr as I had some issues while rebasing this branch... the implementation with most comments resolved can be found in this [link](https://github.com/dhis2-sre/im-web-client/pull/519)

**Pull request closed without merging**